src/python/pants/ox.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import contextlib
+import functools
+import importlib.machinery
+import logging
+import os
+import runpy
+import site
+import sys
+import zipimport
+
+_logger = logging.getLogger(__name__)
+
+
+def bootstrap_pyoxidizer() -> None:
+    if is_oxidized:
+        _logger.info("Pants is running as a PyOxidizer binary.")
+
+
+# Provide the `is_oxidized` symbol, to allow for workarounds in Pants code where we use things
+# that don't work under PyOxidizer's custom importer. `oxidized_importer` is only accessible
+# in Pants under PyOxidizer, so an import failure will occur if we're not oxidized.
+try:
+    import oxidized_importer  # type: ignore # pants: no-infer-dep # noqa: F401
+
+    is_oxidized = True
+except ModuleNotFoundError:
+    is_oxidized = False
+
+
+if is_oxidized and not sys.argv[0]:
+    # A not insignificant amount of Pants code relies on `sys.argv[0]`, which is modified in an
+    # invalid way by python's `pymain_run_module` support. For our purposes, the executable
+    # distribution is the correct `argv[0]`.
+    # See https://github.com/indygreg/PyOxidizer/issues/307
+    sys.argv[0] = sys.executable
+
+
+def pex_main() -> bool:
+    """Detect whether some external process is trying to invoke this binary as Pex.
```

Comment:
IIUC this is because Pants tries to resolve plugins using sys.executable right here:
https://github.com/pantsbuild/pants/blob/d63c8958910dd10c5450ce76563a31030212b6ad/src/python/pants/init/plugin_resolver.py#L46-L70

The ICs are not present via this code path:
https://github.com/pantsbuild/pants/blob/d63c8958910dd10c5450ce76563a31030212b6ad/src/python/pants/init/plugin_resolver.py#L105-L123
https://github.com/pantsbuild/pants/blob/b1864b6e4e45810507aca007fb62a472e28299a1/src/python/pants/init/options_initializer.py#L102

If all this is true, could you just have Pants set an env var or use some other direct signal to tell you here what's up? That would avoid divining tea leaves like this.

Reply:
Or just dont use `sys.executable` ever. Run Pex via a real python until PyOxidizer supports being one of those. That's the IC path, so al the code is there to do this IIUC.

Reply:
We could set an env var for the initial invocation, but the issue at hand is the pile of invocations that are made by Pex in the plugin bootstrapping process (i.e. to run `venv` and `pip` itself), so we'd need to ensure that that specific environment variable is propagated. And we'd still need to look at `argv` to figure out which bootstrapping function to run.

It's worth understanding why we use `sys.executable` in the pre-pyoxidizer case: it's to make sure that plugins are configured with the same Python version and platform as being used to run Pants. We could set ICs and platform constraints by inspecting various things under `sys`, but in an environment where a Python interpreter is present, realistically, it should just end up finding the value of `sys.executable`, just with a giant pile more indirection.

Remember that the goal of this is to support running Pants without the need for a Python interpreter to be present. If we didn't use `sys.executable` and relied on the Python support elsewhere in Pants to load Pex:

* we'd need to have a real Python present
* it would need to be the same version as provided with PyOxidizer, 
* and it'd only be necessary for the purpose of downloading the plugins.
* The external interpreter would never actually run the downloaded packages -- those would be loaded inside Pants

Basically, it would entirely defeat the purpose of not needing a Python interpreter

A better solution would be to not fetch plugins by pip at all, and then as long as we can unzip a wheel file into a reliable location, we wouldn't need to invoke Pex and venv at all. I'd be willing to talk about this use case for a future version of Pants; I think it might make a bit of sense.

Reply:
Ah yes, plugin Python need to match our provided Python in the form of the pyoxidezed one. Gotcha.

This also could've been achieved by just having the wrapper script download a statically linked Python, which I believe the py03 family of projects provides. Then all the problems are solved - we have a real python interpreter to use, the user need not have one installed, etc. The loss is slightly more complexity in our pants wrapper script / setup, but IIUC that script still lives in this world anyhow.

> A better solution would be to not fetch plugins by pip at all, and then as long as we can unzip a wheel file into a reliable location, we wouldn't need to invoke Pex and venv at all. I'd be willing to talk about this use case for a future version of Pants; I think it might make a bit of sense.

That sounds impossible. So would you just require any plugin has to provide some sort of manifest of its transitive dependency wheels, or provides all those wheels too or just outlaw 3rdparty deps for plugins? The latter seems the only workable thing in that world; otherwise Pants needs to reinvent conflict resolution between 3rdparty wheel deps.

Reply:
Maybe we could switch to depending on Pip though and using runpy to execute it?

Reply:
Probably similar issues. It invokes itself in chroots to do sdist builds IIUC.

Reply:
@jsirois The trick with changing the embed approach here is that we need to be able to swap the import machinery back into the non-`pyO3` versions when we're loading Pex, which is what the `use_traditional_import_machinery` decorator does. I think we'd end up with a runner script that looks a lot like `ox.py` in the end.

Reply:
(PyO3 uses a custom importer that loads the stdlib and the packaged python code from strings interned into the binary, but also subsumes the filesystem loading as well. It means that anything that gets loaded by that importer does not have `__file__` set, and `pex` and `venv` both rely heavily on `__file__` being set. )

Reply:
No, the idea I'm proposing doesn't use pyoxidizer at all. It ships Pants as-is. The only difference is the wrapper script doesn't look for Python on the local system, it downloads a pre-compiled one with no Pants in it at all.

Reply:
Right, the PyO3-distributed interpreter has the issue I just described

Reply:
Aha, wow.

Reply:
The custom interpreter is [measurably faster](https://pyoxidizer.readthedocs.io/en/stable/oxidized_importer.html), so it's good that we have it for the most part. It's just that it takes a bit more of a scorched-earth approach to `__file__` than is absolutely necessary. 🤷

Reply:
FWIW It actually looks like the pyoxy interpreter handles `__file__` just fine and this is just using it stock and not customizing it with a yaml config to turn off the pyoxidized importer for the stdlib stuff:
```
jsirois@Gill-Windows:/mnt/c/Users/John Sirois/Downloads/pyoxy-0.2.0-x86_64-unknown-linux-gnu-cpython3.10$ cat package/module.py
import sys


print(__file__, file=sys.stderr)

try:
    print(sys.__file__, file=sys.stderr)
except AttributeError:
    print(f"{sys.meta_path}", file=sys.stderr)

jsirois@Gill-Windows:/mnt/c/Users/John Sirois/Downloads/pyoxy-0.2.0-x86_64-unknown-linux-gnu-cpython3.10$ ./pyoxy run-python -- -m package.module
/mnt/c/Users/John Sirois/Downloads/pyoxy-0.2.0-x86_64-unknown-linux-gnu-cpython3.10/package/module.py
[<oxidized_importer.OxidizedFinder object at 0x7f3e425a7be0>, <class '_frozen_importlib_external.PathFinder'>]
```

I'm realizing I can at least use this over in Pex to support Par I think.